File: DevEnvironment/DataStructures/src/test/java/datastructures/binarytree/core/BinaryTreeTestData.java
package datastructures.binarytree.core;

import base.core.ArrayLists;
import datastructures.binarytree.interfaces.IBinaryTreeData;
import datastructures.binarytree.interfaces.IBinaryTreeNode;
import datastructures.binarytree.interfaces.IBinaryTreeTestData;
import java.util.ArrayList;
import java.util.List;

/**
 * The BinaryTreeTestData class implements data of tests for a binary tree.
 */
public final class BinaryTreeTestData implements IBinaryTreeTestData {
    /**
     * The BinaryTreeTestData constructor.
     */
    public BinaryTreeTestData() {
    }

    /**
     * Gets the data of the trees.
     */
    @Override
    public List<IBinaryTreeData<Integer, String>> getData() {
        List<IBinaryTreeData<Integer, String>> data = new ArrayList<>();

        data.add(this.getTreeData1());
        data.add(this.getTreeData2());
        data.add(this.getTreeData3());
        data.add(this.getTreeData4());

        return data;
    }

    /**
     * Gets the data of tree1.
     */
    private IBinaryTreeData<Integer, String> getTreeData1() {
        List<IBinaryTreeNode<Integer, String>> creationData = new ArrayList<>();
        List<IBinaryTreeNode<Integer, String>> levelOrder = new ArrayList<>();
        List<IBinaryTreeNode<Integer, String>> inorder = new ArrayList<>();
        List<IBinaryTreeNode<Integer, String>> preorder = new ArrayList<>();
        List<IBinaryTreeNode<Integer, String>> postorder = new ArrayList<>();

        return new BinaryTreeData<>(
            creationData,
            levelOrder,
            inorder,
            preorder,
            postorder);
    }

    /**
     * Gets the data of tree2.
     */
    private IBinaryTreeData<Integer, String> getTreeData2() {
        List<IBinaryTreeNode<Integer, String>> creationData = ArrayLists.make(BinaryTreeNodes.make(1, "a"));
        List<IBinaryTreeNode<Integer, String>> levelOrder = ArrayLists.make(BinaryTreeNodes.make(1, "a"));
        List<IBinaryTreeNode<Integer, String>> inorder = ArrayLists.make(BinaryTreeNodes.make(1, "a"));
        List<IBinaryTreeNode<Integer, String>> preorder = ArrayLists.make(BinaryTreeNodes.make(1, "a"));
        List<IBinaryTreeNode<Integer, String>> postorder = ArrayLists.make(BinaryTreeNodes.make(1, "a"));

        return new BinaryTreeData<>(
            creationData,
            levelOrder,
            inorder,
            preorder,
            postorder);
    }

    /**
     * Gets the data of tree3.
     */
    private IBinaryTreeData<Integer, String> getTreeData3() {
        List<IBinaryTreeNode<Integer, String>> creationData = new ArrayList<>();

        creationData.add(BinaryTreeNodes.make(1, "a"));
        creationData.add(BinaryTreeNodes.make(2, "b"));
        creationData.add(BinaryTreeNodes.make(3, "c"));
        creationData.add(BinaryTreeNodes.endNode());
        creationData.add(BinaryTreeNodes.endNode());
        creationData.add(BinaryTreeNodes.make(4, "d"));
        creationData.add(BinaryTreeNodes.endNode());
        creationData.add(BinaryTreeNodes.endNode());
        creationData.add(BinaryTreeNodes.make(5, "e"));
        creationData.add(BinaryTreeNodes.endNode());
        creationData.add(BinaryTreeNodes.endNode());

        List<IBinaryTreeNode<Integer, String>> levelOrder = ArrayLists.make(
            BinaryTreeNodes.make(1, "a"),
            BinaryTreeNodes.make(2, "b"),
            BinaryTreeNodes.make(5, "e"),
            BinaryTreeNodes.make(3, "c"),
            BinaryTreeNodes.make(4, "d"));

        List<IBinaryTreeNode<Integer, String>> inorder = ArrayLists.make(
            BinaryTreeNodes.make(3, "c"),
            BinaryTreeNodes.make(2, "b"),
            BinaryTreeNodes.make(4, "d"),
            BinaryTreeNodes.make(1, "a"),
            BinaryTreeNodes.make(5, "e"));

        List<IBinaryTreeNode<Integer, String>> preorder = ArrayLists.make(
            BinaryTreeNodes.make(1, "a"),
            BinaryTreeNodes.make(2, "b"),
            BinaryTreeNodes.make(3, "c"),
            BinaryTreeNodes.make(4, "d"),
            BinaryTreeNodes.make(5, "e"));

        List<IBinaryTreeNode<Integer, String>> postorder = ArrayLists.make(
            BinaryTreeNodes.make(3, "c"),
            BinaryTreeNodes.make(4, "d"),
            BinaryTreeNodes.make(2, "b"),
            BinaryTreeNodes.make(5, "e"),
            BinaryTreeNodes.make(1, "a"));

        return new BinaryTreeData<>(
            creationData,
            levelOrder,
            inorder,
            preorder,
            postorder);
    }

    /**
     * Gets the data of tree4.
     */
    private IBinaryTreeData<Integer, String> getTreeData4() {
        List<IBinaryTreeNode<Integer, String>> creationData = new ArrayList<>();

        creationData.add(BinaryTreeNodes.make(1, "a"));
        creationData.add(BinaryTreeNodes.make(2, "b"));
        creationData.add(BinaryTreeNodes.make(4, "d"));
        creationData.add(BinaryTreeNodes.endNode());
        creationData.add(BinaryTreeNodes.endNode());
        creationData.add(BinaryTreeNodes.make(5, "e"));
        creationData.add(BinaryTreeNodes.endNode());
        creationData.add(BinaryTreeNodes.endNode());
        creationData.add(BinaryTreeNodes.make(3, "c"));
        creationData.add(BinaryTreeNodes.make(6, "f"));
        creationData.add(BinaryTreeNodes.endNode());
        creationData.add(BinaryTreeNodes.endNode());
        creationData.add(BinaryTreeNodes.make(7, "h"));
        creationData.add(BinaryTreeNodes.endNode());
        creationData.add(BinaryTreeNodes.endNode());

        List<IBinaryTreeNode<Integer, String>> levelOrder = ArrayLists.make(
            BinaryTreeNodes.make(1, "a"),
            BinaryTreeNodes.make(2, "b"),
            BinaryTreeNodes.make(3, "c"),
            BinaryTreeNodes.make(4, "d"),
            BinaryTreeNodes.make(5, "e"),
            BinaryTreeNodes.make(6, "f"),
            BinaryTreeNodes.make(7, "h"));

        List<IBinaryTreeNode<Integer, String>> inorder = ArrayLists.make(
            BinaryTreeNodes.make(4, "d"),
            BinaryTreeNodes.make(2, "b"),
            BinaryTreeNodes.make(5, "e"),
            BinaryTreeNodes.make(1, "a"),
            BinaryTreeNodes.make(6, "f"),
            BinaryTreeNodes.make(3, "c"),
            BinaryTreeNodes.make(7, "h"));

        List<IBinaryTreeNode<Integer, String>> preorder = ArrayLists.make(
            BinaryTreeNodes.make(1, "a"),
            BinaryTreeNodes.make(2, "b"),
            BinaryTreeNodes.make(4, "d"),
            BinaryTreeNodes.make(5, "e"),
            BinaryTreeNodes.make(3, "c"),
            BinaryTreeNodes.make(6, "f"),
            BinaryTreeNodes.make(7, "h"));

        List<IBinaryTreeNode<Integer, String>> postorder = ArrayLists.make(
            BinaryTreeNodes.make(4, "d"),
            BinaryTreeNodes.make(5, "e"),
            BinaryTreeNodes.make(2, "b"),
            BinaryTreeNodes.make(6, "f"),
            BinaryTreeNodes.make(7, "h"),
            BinaryTreeNodes.make(3, "c"),
            BinaryTreeNodes.make(1, "a"));

        return new BinaryTreeData<>(
            creationData,
            levelOrder,
            inorder,
            preorder,
            postorder);
    }
}

File: DevEnvironmentCLion/DataStructuresTest/binarytree/core/BinaryTreeTestData.cpp
#include "PreCompiled.h"

#include "BinaryTreeTestData.h"
#include "BinaryTreeData.h"
#include "BinaryTreeNodes.h"

using namespace test::datastructures::binarytree;

/**
 * The BinaryTreeTestData constructor.
 */
BinaryTreeTestData::BinaryTreeTestData()
{
}

/**
 * The BinaryTreeTestData constructor.
 */
BinaryTreeTestData::~BinaryTreeTestData()
{
}

/**
 * Gets the data of the trees.
 */
std::vector<BinaryTreeDataSharedPtr<int, std::string>> BinaryTreeTestData::getData() const
{
    std::vector<BinaryTreeDataSharedPtr<int, std::string>> data;

    data.push_back(getTreeData1());
    data.push_back(getTreeData2());
    data.push_back(getTreeData3());
    data.push_back(getTreeData4());

    return data;
}

/**
 * Gets the data of tree1.
 */
BinaryTreeDataSharedPtr<int, std::string> BinaryTreeTestData::getTreeData1() const
{
    std::vector<IBinaryTreeNodeSharedPtr<int, std::string>> creationData;
    std::vector<IBinaryTreeNodeSharedPtr<int, std::string>> levelOrder;
    std::vector<IBinaryTreeNodeSharedPtr<int, std::string>> inorder;
    std::vector<IBinaryTreeNodeSharedPtr<int, std::string>> preorder;
    std::vector<IBinaryTreeNodeSharedPtr<int, std::string>> postorder;

    return BinaryTreeDataSharedPtr<int, std::string>(
        new BinaryTreeData<int, std::string>(
            creationData,
            levelOrder,
            inorder,
            preorder,
            postorder));
}

/**
 * Gets the data of tree2.
 */
BinaryTreeDataSharedPtr<int, std::string> BinaryTreeTestData::getTreeData2() const
{
    std::vector<IBinaryTreeNodeSharedPtr<int, std::string>> creationData = {BinaryTreeNodes::make<int, std::string>(1, "a")};
    std::vector<IBinaryTreeNodeSharedPtr<int, std::string>> levelOrder = {BinaryTreeNodes::make<int, std::string>(1, "a")};
    std::vector<IBinaryTreeNodeSharedPtr<int, std::string>> inorder = {BinaryTreeNodes::make<int, std::string>(1, "a")};
    std::vector<IBinaryTreeNodeSharedPtr<int, std::string>> preorder = {BinaryTreeNodes::make<int, std::string>(1, "a")};
    std::vector<IBinaryTreeNodeSharedPtr<int, std::string>> postorder = {BinaryTreeNodes::make<int, std::string>(1, "a")};

    return BinaryTreeDataSharedPtr<int, std::string>(
        new BinaryTreeData<int, std::string>(
            creationData,
            levelOrder,
            inorder,
            preorder,
            postorder));
}

/**
 * Gets the data of tree3.
 */
BinaryTreeDataSharedPtr<int, std::string> BinaryTreeTestData::getTreeData3() const
{
    std::vector<IBinaryTreeNodeSharedPtr<int, std::string>> creationData;

    creationData.push_back(BinaryTreeNodes::make<int, std::string>(1, "a"));
    creationData.push_back(BinaryTreeNodes::make<int, std::string>(2, "b"));
    creationData.push_back(BinaryTreeNodes::make<int, std::string>(3, "c"));
    creationData.push_back(BinaryTreeNodes::endNode<int, std::string>());
    creationData.push_back(BinaryTreeNodes::endNode<int, std::string>());
    creationData.push_back(BinaryTreeNodes::make<int, std::string>(4, "d"));
    creationData.push_back(BinaryTreeNodes::endNode<int, std::string>());
    creationData.push_back(BinaryTreeNodes::endNode<int, std::string>());
    creationData.push_back(BinaryTreeNodes::make<int, std::string>(5, "e"));
    creationData.push_back(BinaryTreeNodes::endNode<int, std::string>());
    creationData.push_back(BinaryTreeNodes::endNode<int, std::string>());

    std::vector<IBinaryTreeNodeSharedPtr<int, std::string>> levelOrder =
    {
        BinaryTreeNodes::make<int, std::string>(1, "a"),
        BinaryTreeNodes::make<int, std::string>(2, "b"),
        BinaryTreeNodes::make<int, std::string>(5, "e"),
        BinaryTreeNodes::make<int, std::string>(3, "c"),
        BinaryTreeNodes::make<int, std::string>(4, "d")
    };

    std::vector<IBinaryTreeNodeSharedPtr<int, std::string>> inorder =
    {
        BinaryTreeNodes::make<int, std::string>(3, "c"),
        BinaryTreeNodes::make<int, std::string>(2, "b"),
        BinaryTreeNodes::make<int, std::string>(4, "d"),
        BinaryTreeNodes::make<int, std::string>(1, "a"),
        BinaryTreeNodes::make<int, std::string>(5, "e")
    };

    std::vector<IBinaryTreeNodeSharedPtr<int, std::string>> preorder =
    {
        BinaryTreeNodes::make<int, std::string>(1, "a"),
        BinaryTreeNodes::make<int, std::string>(2, "b"),
        BinaryTreeNodes::make<int, std::string>(3, "c"),
        BinaryTreeNodes::make<int, std::string>(4, "d"),
        BinaryTreeNodes::make<int, std::string>(5, "e")
    };

    std::vector<IBinaryTreeNodeSharedPtr<int, std::string>> postorder =
    {
        BinaryTreeNodes::make<int, std::string>(3, "c"),
        BinaryTreeNodes::make<int, std::string>(4, "d"),
        BinaryTreeNodes::make<int, std::string>(2, "b"),
        BinaryTreeNodes::make<int, std::string>(5, "e"),
        BinaryTreeNodes::make<int, std::string>(1, "a")
    };

    return BinaryTreeDataSharedPtr<int, std::string>(
        new BinaryTreeData<int, std::string>(
            creationData,
            levelOrder,
            inorder,
            preorder,
            postorder));
}

/**
 * Gets the data of tree4.
 */
BinaryTreeDataSharedPtr<int, std::string> BinaryTreeTestData::getTreeData4() const
{
    std::vector<IBinaryTreeNodeSharedPtr<int, std::string>> creationData;

    creationData.push_back(BinaryTreeNodes::make<int, std::string>(1, "a"));
    creationData.push_back(BinaryTreeNodes::make<int, std::string>(2, "b"));
    creationData.push_back(BinaryTreeNodes::make<int, std::string>(4, "d"));
    creationData.push_back(BinaryTreeNodes::endNode<int, std::string>());
    creationData.push_back(BinaryTreeNodes::endNode<int, std::string>());
    creationData.push_back(BinaryTreeNodes::make<int, std::string>(5, "e"));
    creationData.push_back(BinaryTreeNodes::endNode<int, std::string>());
    creationData.push_back(BinaryTreeNodes::endNode<int, std::string>());
    creationData.push_back(BinaryTreeNodes::make<int, std::string>(3, "c"));
    creationData.push_back(BinaryTreeNodes::make<int, std::string>(6, "f"));
    creationData.push_back(BinaryTreeNodes::endNode<int, std::string>());
    creationData.push_back(BinaryTreeNodes::endNode<int, std::string>());
    creationData.push_back(BinaryTreeNodes::make<int, std::string>(7, "h"));
    creationData.push_back(BinaryTreeNodes::endNode<int, std::string>());
    creationData.push_back(BinaryTreeNodes::endNode<int, std::string>());

    std::vector<IBinaryTreeNodeSharedPtr<int, std::string>> levelOrder =
    {
        BinaryTreeNodes::make<int, std::string>(1, "a"),
        BinaryTreeNodes::make<int, std::string>(2, "b"),
        BinaryTreeNodes::make<int, std::string>(3, "c"),
        BinaryTreeNodes::make<int, std::string>(4, "d"),
        BinaryTreeNodes::make<int, std::string>(5, "e"),
        BinaryTreeNodes::make<int, std::string>(6, "f"),
        BinaryTreeNodes::make<int, std::string>(7, "h")
    };

    std::vector<IBinaryTreeNodeSharedPtr<int, std::string>> inorder =
    {
        BinaryTreeNodes::make<int, std::string>(4, "d"),
        BinaryTreeNodes::make<int, std::string>(2, "b"),
        BinaryTreeNodes::make<int, std::string>(5, "e"),
        BinaryTreeNodes::make<int, std::string>(1, "a"),
        BinaryTreeNodes::make<int, std::string>(6, "f"),
        BinaryTreeNodes::make<int, std::string>(3, "c"),
        BinaryTreeNodes::make<int, std::string>(7, "h")
    };

    std::vector<IBinaryTreeNodeSharedPtr<int, std::string>> preorder =
    {
        BinaryTreeNodes::make<int, std::string>(1, "a"),
        BinaryTreeNodes::make<int, std::string>(2, "b"),
        BinaryTreeNodes::make<int, std::string>(4, "d"),
        BinaryTreeNodes::make<int, std::string>(5, "e"),
        BinaryTreeNodes::make<int, std::string>(3, "c"),
        BinaryTreeNodes::make<int, std::string>(6, "f"),
        BinaryTreeNodes::make<int, std::string>(7, "h")
    };

    std::vector<IBinaryTreeNodeSharedPtr<int, std::string>> postorder =
    {
        BinaryTreeNodes::make<int, std::string>(4, "d"),
        BinaryTreeNodes::make<int, std::string>(5, "e"),
        BinaryTreeNodes::make<int, std::string>(2, "b"),
        BinaryTreeNodes::make<int, std::string>(6, "f"),
        BinaryTreeNodes::make<int, std::string>(7, "h"),
        BinaryTreeNodes::make<int, std::string>(3, "c"),
        BinaryTreeNodes::make<int, std::string>(1, "a")
    };

    return BinaryTreeDataSharedPtr<int, std::string>(
        new BinaryTreeData<int, std::string>(
            creationData,
            levelOrder,
            inorder,
            preorder,
            postorder));
}
